internal/pkg/vm/actor/builtin/initactor/init.go

```diff
@@ -99,3 +138,190 @@ func (*Impl) GetNetwork(ctx runtime.Runtime) (string, uint8, error) {
+	}
+
+	// create id address
+	actorID := state.assignNewID()
+	idAddr, err := address.NewIDAddress(uint64(actorID))
+	if err != nil {
+		return address.Undef, errors.CodeError(err), errors.FaultErrorWrapf(err, "could not create id address with id %d", actorID)
+	}
+
+	// map id to address and vice versa
+	ctx := context.TODO()
+	state.AddressMap, err = a.setId(ctx, rt.Storage(), state.AddressMap, bytesAddr, actorID)
+	if err != nil {
+		return address.Undef, errors.CodeError(err), errors.FaultErrorWrap(err, "could not save id by address")
+	}
+
+	state.IdMap, err = a.setAddress(ctx, rt.Storage(), state.IdMap, actorID, bytesAddr)
+	if err != nil {
+		return address.Undef, errors.CodeError(err), errors.FaultErrorWrap(err, "could not save addres by id")
+	}
+
+	// write the state
+	err = actor.WriteState(rt, state)
+	if err != nil {
+		return address.Undef, errors.CodeError(err), errors.FaultErrorWrap(err, "could not write actor state")
+	}
+
+	// create new actor keyed by id address
+	err = rt.CreateNewActor(idAddr, codeCID)
+	if err != nil {
+		return address.Undef, errors.CodeError(err), errors.FaultErrorWrapf(err, "could not create actor with address %s", idAddr.String())
+	}
+
+	// send message containing actor's initial balance to construct it with the given params
+	_, _, err = rt.Send(idAddr, types.ConstructorMethodID, rt.Message().Value, params)
+	if err != nil {
+		return address.Undef, errors.CodeError(err), err
+	}
+
+	return idAddr, 0, nil
+}
+
+func (a *Impl) setAddress(ctx context.Context, storage runtime.Storage, idMap cid.Cid, actorID types.Uint64, addr address.Address) (cid.Cid, error) {
+	lookup, err := actor.LoadLookup(ctx, storage, idMap)
+	if err != nil {
+		return cid.Undef, errors.RevertErrorWrapf(err, "could not load lookup for cid: %s", idMap)
+	}
+
+	key, err := a.keyForActorID(actorID)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	err = lookup.Set(ctx, key, addr)
+	if err != nil {
+		return cid.Undef, errors.FaultErrorWrapf(err, "could not set address")
+	}
+
+	return lookup.Commit(ctx)
+}
+
+func (a *Impl) setId(ctx context.Context, storage runtime.Storage, addressMap cid.Cid, addr address.Address, actorID types.Uint64) (cid.Cid, error) {
+	lookup, err := actor.LoadLookup(ctx, storage, addressMap)
+	if err != nil {
+		return cid.Undef, errors.RevertErrorWrapf(err, "could not load lookup for cid: %s", addressMap)
+	}
+
+	err = lookup.Set(ctx, addr.String(), actorID)
+	if err != nil {
+		return cid.Undef, errors.FaultErrorWrapf(err, "could not set id")
+	}
+
+	return lookup.Commit(ctx)
+}
+
+func (a *Impl) keyForActorID(actorID types.Uint64) (string, error) {
+	key, err := encoding.Encode(actorID)
+	if err != nil {
+		return "", errors.FaultErrorWrapf(err, "could not encode actor id: %d", actorID)
+	}
+
+	return string(key), nil
+}
+
+func (a *Impl) isBuiltinActor(code cid.Cid) bool {
+	return code.Equals(types.StorageMarketActorCodeCid) ||
```

Comment:
This is awful but in the spec. This should really be available from the context and derived from BuiltinActor data. Also these checks are unnecessary if only non-singleton, builtin actors have constructor methods.

Reply:
Power actor could go in here too